geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java

```diff
@@ -316,10 +323,25 @@ public long rpush(ExecutionHandlerContext context, List<byte[]> elementsToAdd, R
     RedisList sourceList = regionProvider.getTypedRedisData(REDIS_LIST, source, false);
     RedisList destinationList = regionProvider.getTypedRedisData(REDIS_LIST, destination, false);
     Region<RedisKey, RedisData> region = regionProvider.getDataRegion();
-    byte[] moved = sourceList.rpop(region, source);
+
+    if (!sourceList.exists()) {
+      return null;
+    }
+
+    if (source.equals(destination)) {
+      RedisList newSourceList = new RedisList(sourceList);
+      byte[] moved = newSourceList.rpop(region, source);
+      newSourceList.lpush(context, Collections.singletonList(moved), destination, false);
+      return moved;
+    }
+
+    RedisList newSourceList = new RedisList(sourceList);
+    byte[] moved = newSourceList.rpop(region, source);
     if (moved != null) {
-      destinationList.lpush(context, Collections.singletonList(moved), destination, false);
+      RedisList newDestinationList = new RedisList(destinationList);
+      newDestinationList.lpush(context, Collections.singletonList(moved), destination, false);
     }
+
     return moved;
```

Comment:
This block of code can be simplified and made a little more efficient by changing the order in which we're doing things a little:
```
    RedisList sourceList = regionProvider.getTypedRedisData(REDIS_LIST, source, false);

    if (sourceList.isNull()) {
      return null;
    }
    
    RedisList newSourceList = new RedisList(sourceList);
    Region<RedisKey, RedisData> region = regionProvider.getDataRegion();
    byte[] moved = newSourceList.rpop(region, source);

    if (source.equals(destination)) {
      newSourceList.lpush(context, Collections.singletonList(moved), destination, false);
    } else {
      RedisList destinationList = regionProvider.getTypedRedisData(REDIS_LIST, destination, false);
      new RedisList(destinationList).lpush(context, Collections.singletonList(moved), destination, false);
    }
    return moved;
```

By moving the check for whether the source list is null higher up, we can return early without having to retrieve the destination list, which also fixes some incorrect behaviour where a nonexistent source key with a non-list destination key would return an error instead of returning null. An integration test to show this behaviour is:
```
  @Test
  public void rPopLPush_withNonexistentSourceKeyAndNonListDestinationKey_returnsNull() {
    jedis.set(DESTINATION_KEY, "not_a_list");

    assertThat(jedis.rpoplpush(SOURCE_KEY, DESTINATION_KEY)).isNull();
  }
```

We can also avoid having to retrieve the destination list from the region if the source and destination are the same by moving the `getTypedRedisData()` call inside the if statement, which will improve performance in that case. We also don't need to check both that the source is not null and that the member popped is not null, as the only time the member could be null would be if the source was null.